### PR TITLE
New lint [`compressable_if`]

### PR DESCRIPTION
**Pull request by SeseMueller**
We all started somewhere as programmers. Nowhere is that more appearent than looking back at what code you wrote when you knew less and cringing because it was *just that bad*.

A very popular example, as many know, is the problem of determining whether a number is even or odd. 
Knowing nothing about the modulus operator or matching in rust, a beginner might write:
``` rust
if test == 1 {
        println!("odd");
    }
    if test == 2 {
        println!("even");
    }
    if test == 3 {
        println!("odd");
    }
```

This, however, is inefficient and takes a while to run. 

However, with the power of clippy and 🚀blazingly fast lints ✨, we can help.

Introducing:
### The `compressable_if` lint!
(Not to be confused with the `collapsable_if` lint)

It automatically detects if two if-blocks can be merged if their inner blocks are identical and suggesst merging them via OR-ing the conditions.

This allows even beginners to write much more concise code!
For example, the above example becomes 

```rust
if test == 1 || test == 3 {
        println!("odd");
    }
    if test == 2 {
        println!("even");
    }
```
and with *the power of clippy* they can even improve their code with this lint **with a single click!**

All in all, this lint will help beginners understand how to write better code and will totally not lead to any code breakage whatsoever.

For further info, read here:
<details><summary>Details</summary>
<p>

First of all, I want to make clear that this is an april fools joke. 
The lint is sloppy and has many false positives and even breaks code from time to time. 

It was hard to even get it so pass because it hit so many times on clippy internals, because many tests test against empty if blocks. (Which is also the reason why 156 files were changed)

I first had the idea when the even-odd craze came over many programming social media channels. 
Originally, I wanted the lint to output code with the modulus operator, but I found it a bit too hard given I have little experience and not much time.

I hardly doubt that an actual `collapsable_if` is possible/reasonable, since the nessecary logic is extremely complex and could lead to incomprehensible conditions. I certainly won't try.

Happy april fools!
Stay safe and don't use lint from people you don't trust.

(Also I really hope that the check passes; I worked quite hard on it.)

</p>
</details>

### CODE REVIEW

**Comment by matthiaskrgr** [comment]
thanks!
@​bors r+

**Comment by xFrednet** [comment]
@matthiaskrgr was this r+ intentional?

**Comment by xFrednet** [comment]
The examples is looks pretty verbose. Do we maybe want to suggest to use shorter identifies?

```rs
    if t == 1 || t == 3 {
        println!("odd");
    }
    if t == 2 {
        println!("even");
    }
```

Imagine how much disk space we could safe!

File: clippy_lints/src/bool_assert_comparison.rs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#![allow(clippy::compressable_if)]
```
Comment (llogiq) [inline clippy_lints/src/bool_assert_comparison.rs:1]: I'm a bit wary to add a complexity lint that needs multiple `allow` annotations *in our own code*.

I'd either see the lint applied everywhere we can (which would show if it actually is an improvement), or keep it in nursery or pedantic for now.

The last thing we need is a lint that will trigger a lot on average code with a warning that has no urgency at all.

**Comment by J-ZhengLi** [comment]
> The examples is looks pretty verbose. Do we maybe want to suggest to use shorter identifies?
> 
> ```rust
>     if t == 1 || t == 3 {
>         println!("odd");
>     }
>     if t == 2 {
>         println!("even");
>     }
> ```
> 
> Imagine how much disk space we could safe!

after [this PR](https://github.com/rust-lang/rust/pull/123325) is merged, we can save even more unnecessary disk usage! 😆 



But anyway, in case letting more people confused by this April fools PR (especially there's an actual `stacked_ifs` PR exists), I'm gonna close this hope you don't mind~

Thank you for the effort tho, lol

**Comment by SeseMueller** [comment]
Closing it is a good idea imo, I made sure that it was kinda obvious that this is an april fools joke, the other issue was more subtle. Hopefully people look into the issue a bit deeper when they wonder why it's closed vs if it just looks like it's abandoned.
(Also I haven't abandoned the other lint I was working on; it was not just warming up for this one, but the semester has begun and I'll get back to that one later)

**Pull request closed without merging**